### PR TITLE
JsonDB functionality. Program should now save to Json at exit and load from Json on startup.

### PR DESCRIPTION
**Pull request by Kviick**

### CODE REVIEW

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+
+    /**
+     * addToJson() automatically adds an Object to the corresponding json file without erasing its contents, if it exists.
+     *
+     * @param objectToAdd
+     * @throws IOException
+     */
+    public static void addToJson(Object objectToAdd) throws IOException {
+        String fileName;
+        if (objectToAdd instanceof Item) {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: instanceof är lite fult, men kanske måste kanske användas. Går det åtminstone att byta ut till interfaces?

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    /**
+     * addListToJson() adds a list of choice to a json file of choice without erasing the contents that are already in the json file.
+     *
+     * @param listToAdd
+     * @throws IOException
+     */
+
+    public static void addListToJson(List listToAdd) throws IOException {
+        try {
+            if (listToAdd.get(0) instanceof Item) {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: Same as above, is instsanseof necessary+

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+    /**
+     * Uses private methods to get a list.
+     *
+     * @param typeOfList
+     * @return a list with objects from a json file corresponding to typeOfList
+     * @throws IOException
+     */
+
+    public static List getListFromJson(Type typeOfList) throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: Is it possible to set type for List? Unsure how this works

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        } else {
+            throw new IllegalArgumentException("There is no JSON file for chosen type of list!");
+        }
+    }
+
+    /**
+     * @return a list of Items from the itemDB file
+     * @throws IOException
+     */
+    private static List getItemList() throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: List should have a type

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            itemList.add(gson.fromJson(o.toString(), Item.class));
+        }
+        return itemList;
+    }
+
+    /**
+     * @return a list of Locations from the locationDB file
+     * @throws IOException
+     */
+    private static List getLocationList() throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: List should have a type

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+    /**
+     * Writes an Object to the json file corresponding to fileName
+     *
+     * @param objectToAdd
+     * @param fileName
+     * @throws IOException
+     */
+    private static void addToJson(Object objectToAdd, String fileName) throws IOException {
+        GsonBuilder gsonBuilder = Converters.registerInterval(new GsonBuilder());
+        gsonBuilder.registerTypeAdapter(IBorrower.class, new BorrowerSerializer());
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: Don't know Gson, but this looks like duplication to me and to intellij

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        return locationList;
+    }
+
+    /**
+     * Handles IBorrower, IReservable and IReservation interface types through BorrowerSerialiser, ReservableSerialiser and ReservationSerializer classes. Handles org.joda.time.Interval through Converters.registerInterval()
+     *
+     * @return a list of Organisations from the organisationDB file
+     * @throws IOException
+     */
+    private static List getOrganisationList() throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: List should have a type

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    }
+
+    /**
+     * Handles IBorrower, IReservable and IReservation interface types through BorrowerSerialiser, ReservableSerialiser and ReservationSerializer classes. Handles org.joda.time.Interval through Converters.registerInterval()
+     *
+     * @return a list of Reservations from the reservationDB file
+     * @throws IOException
+     */
+
+    private static List getReservationList() throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: List should have type

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        }
+        return reservationList;
+    }
+
+    /**
+     * @return a list of Teams from the teamDB file
+     * @throws IOException
+     */
+
+    private static List getTeamList() throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: List should have Type

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        }
+        return teamList;
+    }
+
+    /**
+     * @return a list of Users from the userDB file
+     * @throws IOException
+     */
+
+    private static List getUserList() throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: List should have type

File: src/main/java/storagesystem/services/ReservableSerializer.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
+package storagesystem.services;
+
+import com.google.gson.*;
+import storagesystem.model.IReservable;
+
+import java.lang.reflect.Type;
+
+public class ReservableSerializer implements JsonSerializer<IReservable>, JsonDeserializer<IReservable> {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/ReservableSerializer.java]: Javadoc

File: src/main/java/storagesystem/services/ReservationSerializer.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
+package storagesystem.services;
+
+import com.google.gson.*;
+import storagesystem.model.IBorrower;
+import storagesystem.model.IReservable;
+import storagesystem.model.IReservation;
+
+import java.lang.reflect.Type;
+
+public class ReservationSerializer implements JsonSerializer<IReservation>, JsonDeserializer<IReservation> {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/ReservationSerializer.java]: Javadoc

File: src/main/java/storagesystem/services/ReservationSerializer.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+    public JsonElement serialize(IReservation jsonElement, Type type, JsonSerializationContext jsonSerializationContext) {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty(CLASSNAME, jsonElement.getClass().getName());
+        jsonObject.add(DATA, jsonSerializationContext.serialize(jsonElement));
+        return jsonObject;
+    }
+
+    /****** Helper method to get the className of the object to be deserialized *****/
+    public Class getObjectClass(String className) {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/ReservationSerializer.java]: Can be private

File: src/main/java/storagesystem/model/Item.java [hunk 14]
```diff
@@ -23,10 +24,10 @@
     private int amount;
     private Condition condition;
     private boolean reservable;
-    private Location location;
-    private Image image;
+    private int locationID;
```
Comment (Sponken) [inline src/main/java/storagesystem/model/Item.java]: Is this change good?

File: src/main/java/storagesystem/viewcontroller/itemview/ItemPageController.java [hunk 15]
```diff
@@ -80,8 +81,8 @@ private void updateAllVisibleFields() {
         setConditionSlider(item.getCondition());
         setReservableLabel(item.isReservable() + "");
         setReservableBtn(item.isReservable());
-        setLocationLabel(item.getLocation().getName());
-        setImage(item.getImage());
+        setLocationLabel(StoreIT.getCurrentOrganisation().getLocation(item.getLocationID()).getName());
```
Comment (Sponken) [inline src/main/java/storagesystem/viewcontroller/itemview/ItemPageController.java]: Changing location in item to locationID makes this more complicated

File: src/test/java/storagesystem/model/ReservationHandlerTest.java [hunk 16]
```diff
@@ -120,3 +120,5 @@ public void shouldReturnAllBorrowerReservations() {
     }
 
 }
+
+ */
```
Comment (Sponken) [inline src/test/java/storagesystem/model/ReservationHandlerTest.java]: Why is this commented???

File: src/test/java/storagesystem/model/ReservationTest.java [hunk 17]
```diff
@@ -42,3 +42,5 @@ public void objectShouldEqual(){
     }
 
 }
+
+ */
```
Comment (Sponken) [inline src/test/java/storagesystem/model/ReservationTest.java]: Why is this outcommented?

File: src/main/java/storagesystem/model/Organisation.java [hunk 18]
```diff
@@ -87,6 +141,18 @@ IReservable getItem(int ID) throws NoSuchElementException {
         return usersTeams;
     }
 
+    public List<IReservable> getTeamsItems(Team team) {
+        List<IReservable> teamsItems = new ArrayList<>();
+        for (IReservable item : items) {
+            for (int itemID : team.getAllItemIDs()) {
```
[the remark this answers is not in the document]
Comment (Kviick) [reply src/main/java/storagesystem/model/Organisation.java]: Optimization is the root to all evil

File: src/main/java/storagesystem/model/StoreIT.java [hunk 19]
```diff
@@ -17,17 +19,44 @@
     /**
      * Loads all data into the program. Should be run at start.
      */
-    public void initializeBackend() {
-        mockData();
+    public void initializeBackend() throws IOException {
+        //JSONHandler.clearAllJsonFiles();
+        //mockData();
```
[the remark this answers is not in the document]
Comment (Kviick) [reply src/main/java/storagesystem/model/StoreIT.java]: No, they're used to reset the application after tests

File: src/main/java/storagesystem/services/JSONHandler.java (hunk 1 above)
Comment (Kviick) [reply src/main/java/storagesystem/services/JSONHandler.java]: Har försökt, lyckas inte lösa det på något bra sätt då fileName måste vara annorlunda beroende på vilken typ som sparas.

File: src/main/java/storagesystem/services/JSONHandler.java (hunk 3 above)
Comment (Kviick) [reply src/main/java/storagesystem/services/JSONHandler.java]: For item we use getListFromJson(Item.class) and get a List containing Items.

File: src/main/java/storagesystem/model/StoreIT.java [hunk 20]
```diff
@@ -17,18 +19,43 @@
...
+            System.out.println("Organisation json is empty.");
+        }
+        System.out.println("Loaded Organisations from json.");
+        IDHandler.updateAllIDs(organisations);
+        currentOrganisation = organisations.get(0);
+        System.out.println("Current Organisation Set.");
     }
 
-    private static void mockData() {
+    private static void mockData() throws IOException {
```
Comment (steget) [inline src/main/java/storagesystem/model/StoreIT.java]: maybe just clear all json files when this method is run?

File: src/main/java/storagesystem/model/Team.java [hunk 21]
```diff
@@ -9,22 +12,36 @@
...
 
     public Team(String teamName) {
         this.name = teamName;
-        //todo fill stuff from db
         termsAndConditions = "";
         teamID = nextID;
         nextID++;
     }
 
+    public void addUsers() throws IOException {
```
Comment (steget) [inline src/main/java/storagesystem/model/Team.java]: call from constructor?

File: src/main/java/storagesystem/services/BorrowerSerializer.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public JsonElement serialize(IBorrower jsonElement, Type type, JsonSerializationContext jsonSerializationContext) {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty(CLASSNAME, jsonElement.getClass().getName());
+        jsonObject.add(DATA, jsonSerializationContext.serialize(jsonElement));
+        return jsonObject;
+    }
+
+    /****** Helper method to get the className of the object to be deserialized *****/
```
Comment (steget) [inline src/main/java/storagesystem/services/BorrowerSerializer.java]: not correct javadoc format

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+            }
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("List is empty!");
+        }
+    }
+
+    /**
+     * @param listToAdd
+     * @param database
+     * @throws IOException
```
Comment (steget) [inline src/main/java/storagesystem/services/JSONHandler.java]: some explanations maybe?

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+     */
+
+    public static void clearJson(String fileName) throws IOException {
+        Writer writer = new FileWriter(fileName);
+        writer.flush();
+        writer.close();
+    }
+
+    /**
+     * clearAllJsonFiles() clears all json files declared completely.
```
Comment (steget) [inline src/main/java/storagesystem/services/JSONHandler.java]: declared completely? Weird sentence structure

File: src/main/java/storagesystem/services/ReservationSerializer.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public JsonElement serialize(IReservation jsonElement, Type type, JsonSerializationContext jsonSerializationContext) {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty(CLASSNAME, jsonElement.getClass().getName());
+        jsonObject.add(DATA, jsonSerializationContext.serialize(jsonElement));
+        return jsonObject;
+    }
+
+    /****** Helper method to get the className of the object to be deserialized *****/
```
Comment (steget) [inline src/main/java/storagesystem/services/ReservationSerializer.java]: remove some stars

File: src/test/java/storagesystem/model/OrganisationTest.java [hunk 26]
```diff
@@ -21,28 +27,39 @@ public void nameTest() {
...
 
     @Test(expected = Exception.class)
     public void getItemTest() throws Exception {
+        JSONHandler.clearAllJsonFiles();
+
         Organisation informationsteknik = new Organisation("Informationsteknik");
         Team tempTeam = new Team("sexNollK");
         informationsteknik.addTeam(tempTeam);
-        IReservable mockItem = IReservableFactory.createReservableItem("mockItem", "This is a description", "Behave please.", 2, Condition.GOOD, true, new Location("testLocation", "test desc", null), null);
+        Location mockLocation = new Location("MockLocation", "this is a mock location in getItemTest()", null);
```
Comment (steget) [inline src/test/java/storagesystem/model/OrganisationTest.java]: don't use null

File: src/test/java/storagesystem/model/ReservationHandlerTest.java [hunk 27]
```diff
@@ -17,8 +17,8 @@
     @Test
     public void shouldCreateMultipleReservations() {
 
-        IBorrower borrower = new User("John Doe", "Developer", "Call me");
-        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
+        IBorrower borrower = new Team("Team1");
+        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
```
Comment (steget) [inline src/test/java/storagesystem/model/ReservationHandlerTest.java]: don't use null

File: src/test/java/storagesystem/model/ReservationHandlerTest.java [hunk 28]
```diff
@@ -40,9 +40,9 @@ public void shouldCreateMultipleReservations() {
 
     @Test
     public void onlyOneReservationPerObjectAndIntervalShouldExist() {
-        IBorrower borrower = new User("John Doe", "Developer", "Call me");
-        IReservable object1 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
-        IReservable object2 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
+        IBorrower borrower = new Team("Team1");
+        IReservable object1 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
+        IReservable object2 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
```
Comment (steget) [inline src/test/java/storagesystem/model/ReservationHandlerTest.java]: don't use null

File: src/test/java/storagesystem/model/ReservationHandlerTest.java [hunk 29]
```diff
@@ -70,8 +70,8 @@ public void onlyOneReservationPerObjectAndIntervalShouldExist() {
 
     @Test
     public void shouldGiveID() {
-        IBorrower borrower = new User("John Doe", "Developer", "Call me");
-        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
+        IBorrower borrower = new Team("Team1");
+        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
```
Comment (steget) [inline src/test/java/storagesystem/model/ReservationHandlerTest.java]: don't use null

File: src/test/java/storagesystem/model/ReservationHandlerTest.java [hunk 30]
```diff
@@ -93,11 +93,11 @@ public void shouldGiveID() {
...
     public void shouldReturnAllBorrowerReservations() {
-        IBorrower borrower1 = new User("John Doe", "Developer", "Call me");
-        IBorrower borrower2 = new User("William", "Admin", "Mail me");
+        IBorrower borrower1 = new Team("Team1");
+        IBorrower borrower2 = new Team("Team1");
 
-        IReservable object1 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
-        IReservable object2 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
+        IReservable object1 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
+        IReservable object2 = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
```
Comment (steget) [inline src/test/java/storagesystem/model/ReservationHandlerTest.java]: don't use null

File: src/test/java/storagesystem/model/ReservationTest.java [hunk 31]
```diff
@@ -9,8 +9,8 @@
 
     @Test
     public void reservationsShouldBeEqual(){
-        IBorrower borrower = new User("John Doe", "Developer", "Call me");
-        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
+        IBorrower borrower = new Team("Team1");
+        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
```
Comment (steget) [inline src/test/java/storagesystem/model/ReservationTest.java]: don't use null

File: src/test/java/storagesystem/model/ReservationTest.java [hunk 32]
```diff
@@ -26,8 +26,8 @@ public void reservationsShouldBeEqual(){
 
     @Test
     public void objectShouldEqual(){
-        IBorrower borrower = new User("John Doe", "Developer", "Call me");
-        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, null, null);
+        IBorrower borrower = new Team("Team1");
+        IReservable object = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, 0, null);
```
Comment (steget) [inline src/test/java/storagesystem/model/ReservationTest.java]: don't use null

File: src/test/java/storagesystem/model/TeamTest.java [hunk 33]
```diff
@@ -2,27 +2,29 @@
...
+        Organisation org = new Organisation("Test org");
         Team testTeam = new Team("owls");
-        List<IReservable> allItems = testTeam.getAllItems();
+        List<IReservable> allItems = org.getTeamsItems(testTeam);
+
         assertEquals(0, allItems.size());
         Location hasen = new Location("hasen","ha senare", null);
 
-        IReservable mockItem = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, hasen, null);
+        IReservable mockItem = IReservableFactory.createReservableItem("mockItem", "desc","requirements",1,Condition.GREAT,true, hasen.getID(), null);
```
Comment (steget) [inline src/test/java/storagesystem/model/TeamTest.java]: don't use null

File: src/main/java/storagesystem/Main.java [hunk 34]
```diff
@@ -30,4 +33,20 @@ public void start(Stage stage) throws Exception {
         stage.setResizable(false);
         stage.show();
     }
+    @Override
+    public void stop() throws IOException {
+        JSONHandler.clearAllJsonFiles();
+        JSONHandler.addListToJson(StoreIT.getOrganisations());
+        System.out.println("Saved Organisations.");
+        JSONHandler.addListToJson(StoreIT.getCurrentOrganisation().getUsers());
```
[the remark this answers is not in the document]
Comment (Kviick) [reply src/main/java/storagesystem/Main.java]: Changed in new Push

File: src/main/java/storagesystem/model/Item.java (hunk 14 above)
Comment (Kviick) [reply src/main/java/storagesystem/model/Item.java]: maybe

File: src/main/java/storagesystem/model/Organisation.java [hunk 35]
```diff
@@ -12,39 +15,73 @@
...
+            this.teams.addAll(teamList);
+        } catch (NullPointerException e) {
+            System.out.println("Team json is empty.");
+        }
+        try {
+            List<User> userList = JSONHandler.getListFromJson(User.class);
+            this.users.addAll(userList);
+        } catch (NullPointerException e) {
+            System.out.println("User json is empty");
+        }
```
[the remark this answers is not in the document]
Comment (Kviick) [reply src/main/java/storagesystem/model/Organisation.java]: Removed ALL other files but organisationDB in new push

File: src/main/java/storagesystem/model/Organisation.java [hunk 36]
```diff
@@ -12,39 +15,73 @@
...
+        locations.addAll(locationList);
+        List<Item> itemList = JSONHandler.getListFromJson(Item.class);
+        items.addAll(itemList);
+        List<Team> teamList = JSONHandler.getListFromJson(Team.class);
+        teams.addAll(teamList);
+        List<User> userList = JSONHandler.getListFromJson(User.class);
+        users.addAll(userList);
         reservationHandler = new ReservationHandler();
-    }
+    }*/
```
[the remark this answers is not in the document]
Comment (Kviick) [reply src/main/java/storagesystem/model/Organisation.java]: Removed as much as I could find

File: src/main/java/storagesystem/model/ReservationHandler.java [hunk 37]
```diff
@@ -23,7 +24,7 @@ public ReservationHandler(List<IReservation> reservations) {
         this.reservations = reservations;
     }
 
-    public ReservationHandler() {
+    public ReservationHandler() throws IOException {
```
[the remark this answers is not in the document]
Comment (Kviick) [reply src/main/java/storagesystem/model/ReservationHandler.java]: Fixed

File: src/main/java/storagesystem/model/StoreIT.java (hunk 20 above)
Comment (Kviick) [reply src/main/java/storagesystem/model/StoreIT.java]: added reset() method instead

File: src/main/java/storagesystem/services/JSONHandler.java (hunk 1 above)
Comment (Kviick) [reply src/main/java/storagesystem/services/JSONHandler.java]: Removed most methods and ALL instanceof in new push

File: src/main/java/storagesystem/viewcontroller/itemview/ItemPageController.java (hunk 15 above)
Comment (Kviick) [reply src/main/java/storagesystem/viewcontroller/itemview/ItemPageController.java]: true. But it's easier to write into json and fewer dependencies in Item?

File: src/main/java/storagesystem/model/StoreIT.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (Kviick) [reply src/main/java/storagesystem/model/StoreIT.java]: added reset() method that may be used?
Or do we remove mockData() alltogether?

File: pom.xml [hunk 38]
```diff
@@ -64,6 +64,12 @@
             <artifactId>gson</artifactId>
             <version>2.8.5</version>
         </dependency>
+
+        <dependency>
+            <groupId>com.fatboyindustrial.gson-jodatime-serialisers</groupId>
+            <artifactId>gson-jodatime-serialisers</artifactId>
```
[the remark this answers is not in the document]
Comment (Kviick) [reply pom.xml]: serialiser used in gson to serialise Interval in JSONHandler

File: src/main/java/storagesystem/model/Organisation.java [hunk 39]
```diff
@@ -21,58 +19,13 @@
     private final List<Team> teams = new ArrayList<>();
     private final List<User> users = new ArrayList<>();
     private final List<Location> locations = new ArrayList<>();
-    private ReservationHandler reservationHandler;
+    private ReservationHandler reservationHandler = new ReservationHandler();
     //todo reservationHandler
```
Comment (steget) [inline src/main/java/storagesystem/model/Organisation.java]: remove this todo 😅

File: src/main/java/storagesystem/model/Team.java [hunk 40]
```diff
@@ -28,15 +25,13 @@ public Team(String teamName) {
         nextID++;
     }
 
-    public void addUsers() throws IOException {
-        List<User> userList = JSONHandler.getListFromJson(User.class);
+    public void addUsers(List<User> userList) {
         for (int i = 0; i < userList.size(); i++) {
```
Comment (steget) [inline src/main/java/storagesystem/model/Team.java]: for each instead?

File: src/main/java/storagesystem/model/Team.java [hunk 41]
```diff
@@ -28,15 +25,13 @@ public Team(String teamName) {
...
             memberIDs.add(userList.get(i).getID());
         }
     }
 
-    public void addItems() throws IOException {
-        List<Item> itemList = JSONHandler.getListFromJson(Item.class);
+    public void addItems(List<IReservable> itemList) {
         for (int i = 0; i < itemList.size(); i++) {
```
Comment (steget) [inline src/main/java/storagesystem/model/Team.java]: for each instead?

File: src/main/java/storagesystem/services/IDHandler.java [hunk 42]
```diff
@@ -2,9 +2,11 @@
 
 import storagesystem.model.*;
 
+import javax.naming.spi.ResolveResult;
```
Comment (steget) [inline src/main/java/storagesystem/services/IDHandler.java]: Not used
```suggestion
```

File: src/main/java/storagesystem/model/IReservable.java [hunk 43]
```diff
@@ -32,9 +32,9 @@
 
     boolean isReservable();
 
-    Location getLocation();
+    int getLocationID();
 
-    Image getImage();
+    String getImageUrl();
```
Comment (Sponken) [inline src/main/java/storagesystem/model/IReservable.java]: This should be removed

File: src/main/java/storagesystem/model/IReservable.java [hunk 44]
```diff
@@ -52,8 +52,8 @@
 
     public void setReservable(boolean reservable);
 
-    public void setLocation(Location location);
+    public void setLocationID(int locationID);
 
-    public void setImage(Image image);
+    public void setImageUrl(String imageUrl);
```
Comment (Sponken) [inline src/main/java/storagesystem/model/IReservable.java]: This should be removed

File: src/main/java/storagesystem/model/Item.java [hunk 45]
```diff
@@ -24,10 +23,10 @@
...
+    private String imageUrl;
```
Comment (Sponken) [inline src/main/java/storagesystem/model/Item.java]: This should be removed

File: src/main/java/storagesystem/model/IReservableFactory.java [hunk 46]
```diff
@@ -12,7 +11,7 @@
     private IReservableFactory() {
     }
 
-    public static IReservable createReservableItem(String name, String description, String userRequirements, int amount, Condition condition, boolean reservable, Location location, Image image) {
-        return new Item(name, description, userRequirements, amount, condition, reservable, location, image);
+    public static IReservable createReservableItem(String name, String description, String userRequirements, int amount, Condition condition, boolean reservable, int locationID, String imageUrl) {
+        return new Item(name, description, userRequirements, amount, condition, reservable, locationID, imageUrl);
```
Comment (Sponken) [inline src/main/java/storagesystem/model/IReservableFactory.java]: ImageUrl

File: src/main/java/storagesystem/model/Organisation.java [hunk 47]
```diff
@@ -12,41 +13,33 @@
  * @author Hugo Stegrell, Pär Aronsson
  */
 public class Organisation {
+    private String name;
+    private String imageUrl;
```
Comment (Sponken) [inline src/main/java/storagesystem/model/Organisation.java]: imageUrl

File: src/main/java/storagesystem/model/Organisation.java [hunk 48]
```diff
@@ -12,41 +13,33 @@
...
-        List<IReservable> allItems = new ArrayList<IReservable>();
-        for (Team t :
-                teams) {
-            allItems.addAll(t.getAllItems());
-        }
-        return allItems;
+        return items;
+    }
+
+    public void addItem(IReservable iReservable, Team team) {
```
Comment (Sponken) [inline src/main/java/storagesystem/model/Organisation.java]: Javadoc

File: src/main/java/storagesystem/model/Organisation.java [hunk 49]
```diff
@@ -100,6 +109,16 @@ public boolean isUserPartOfTeam(IBorrower user){
         return usersTeams;
     }
 
+    public List<IReservable> getTeamsItems(Team team) {
```
Comment (Sponken) [inline src/main/java/storagesystem/model/Organisation.java]: Javadoc

File: src/main/java/storagesystem/model/StoreIT.java [hunk 50]
```diff
@@ -118,7 +121,21 @@ public static void setCurrentTeam(Team currentTeam) {
     /**
      * Loads all data into the program. Should be run at start.
      */
-    public void initializeBackend() {
+    public void initializeBackend() throws IOException {
+        //reset();
```
Comment (Sponken) [inline src/main/java/storagesystem/model/StoreIT.java]: What is this? Remove or comment

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 51]
```diff
@@ -0,0 +1,123 @@
...
+
+import java.io.*;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * @author Carl Lindh
+ * JSONHandler is a handler with static methods with which it is possible to write and read from the json files corresponding to the Strings declared.
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: Not Strings in plural anymore, only Organisation

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 52]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    /**
+     * addListToJson() adds a list of choice to a json file of choice without erasing the contents that are already in the json file.
+     *
+     * @param listToAdd
+     * @throws IOException
+     */
+
+    static void addListToJson(List listToAdd) throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: private?

File: src/main/java/storagesystem/services/JSONHandler.java [hunk 53]
```diff
@@ -0,0 +1,123 @@
...
+        System.out.println("Saved Organisations.");
+    }
+
+    /**
+     * addToJson() adds an Object to the json file without erasing its contents.
+     *
+     * @param objectToAdd
+     * @throws IOException
+     */
+    static void addToJson(Object objectToAdd) throws IOException {
```
Comment (Sponken) [inline src/main/java/storagesystem/services/JSONHandler.java]: private?

File: src/main/java/storagesystem/model/Organisation.java (hunk 47 above)
Comment (Kviick) [reply src/main/java/storagesystem/model/Organisation.java]: Should still be there

File: src/main/java/storagesystem/model/StoreIT.java (hunk 50 above)
Comment (Kviick) [reply src/main/java/storagesystem/model/StoreIT.java]: reset() is commented

**Pull request merged**